File: tests/sdk/nodejs/test/iam.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateAccessKeyCommand,
  CreateRoleCommand,
  CreateUserCommand,
  DeleteAccessKeyCommand,
  DeleteRoleCommand,
  DeleteUserCommand,
  GenerateCredentialReportCommand,
  GetAccessKeyLastUsedCommand,
  GetCredentialReportCommand,
  GetRoleCommand,
  GetUserCommand,
} from "@aws-sdk/client-iam";
import { GetCallerIdentityCommand, STSClient } from "@aws-sdk/client-sts";
import {
  endpoint,
  newIAM,
  region,
  requireReady,
  uniquePrefix,
} from "../lib/helpers.mjs";

const TRUST =
  '{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}';

test("IAM user and role round-trip", async (t) => {
  if (!(await requireReady(t))) return;
  const client = newIAM();
  const prefix = uniquePrefix();
  const userName = `${prefix}-user`;
  const roleName = `${prefix}-role`;

  await client.send(new CreateUserCommand({ UserName: userName }));
  t.after(async () => {
    try {
      await client.send(new DeleteUserCommand({ UserName: userName }));
    } catch {
      /* ignore */
    }
  });

  const user = await client.send(new GetUserCommand({ UserName: userName }));
  assert.equal(user.User?.UserName, userName);

  await client.send(
    new CreateRoleCommand({
      RoleName: roleName,
      AssumeRolePolicyDocument: TRUST,
    }),
  );
  t.after(async () => {
    try {
      await client.send(new DeleteRoleCommand({ RoleName: roleName }));
    } catch {
      /* ignore */
    }
  });

  const role = await client.send(new GetRoleCommand({ RoleName: roleName }));
  assert.ok(role.Role?.Arn, "GetRole missing ARN");

  await client.send(new DeleteRoleCommand({ RoleName: roleName }));
  await client.send(new DeleteUserCommand({ UserName: userName }));
});

test("IAM access key last-used and credential report", async (t) => {
  if (!(await requireReady(t))) return;
  const client = newIAM();
  const userName = `${uniquePrefix()}-lastused`;

  await client.send(new CreateUserCommand({ UserName: userName }));
  t.after(async () => {
    try {
      await client.send(new DeleteUserCommand({ UserName: userName }));
    } catch {
      /* ignore */
    }
  });

  const keyOut = await client.send(
    new CreateAccessKeyCommand({ UserName: userName }),
  );
  const accessKeyId = keyOut.AccessKey?.AccessKeyId;
  const secretAccessKey = keyOut.AccessKey?.SecretAccessKey;
  assert.ok(accessKeyId && secretAccessKey, "CreateAccessKey missing key");
  t.after(async () => {
    try {
      await client.send(
        new DeleteAccessKeyCommand({
          UserName: userName,
          AccessKeyId: accessKeyId,
        }),
      );
    } catch {
      /* ignore */
    }
  });

  const userSTS = new STSClient({
    region: region(),
    endpoint: endpoint(),
    credentials: {
      accessKeyId,
      secretAccessKey,
    },
  });
  await userSTS.send(new GetCallerIdentityCommand({}));

  const lastUsed = await client.send(
    new GetAccessKeyLastUsedCommand({ AccessKeyId: accessKeyId }),
  );
  assert.equal(lastUsed.UserName, userName);
  assert.equal(lastUsed.AccessKeyLastUsed?.ServiceName, "sts");
  assert.equal(lastUsed.AccessKeyLastUsed?.Region, region());

  const gen = await client.send(new GenerateCredentialReportCommand({}));
  assert.equal(String(gen.State || "").toUpperCase(), "COMPLETE");

  const report = await client.send(new GetCredentialReportCommand({}));
  assert.ok(report.Content?.length, "GetCredentialReport empty Content");
  const csvText = Buffer.from(report.Content).toString("utf8");
  const found = csvText
    .split(/\r?\n/)
    .slice(1)
    .some((line) => line.split(",")[0] === userName);
  assert.ok(found, `credential report missing user ${userName}: ${csvText}`);
});

File: tests/sdk/go/iam_test.go
package sdk_test

import (
	"context"
	"encoding/csv"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/iam"
	"github.com/aws/aws-sdk-go-v2/service/sts"
)

func TestIAMUserAndRoleRoundTrip(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newIAM(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)

	userName := prefix + "-user"
	_, err := client.CreateUser(ctx, &iam.CreateUserInput{UserName: aws.String(userName)})
	if err != nil {
		t.Fatalf("CreateUser: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteUser(ctx, &iam.DeleteUserInput{UserName: aws.String(userName)})
	})

	user, err := client.GetUser(ctx, &iam.GetUserInput{UserName: aws.String(userName)})
	if err != nil {
		t.Fatalf("GetUser: %v", err)
	}
	if user.User == nil || user.User.UserName == nil || *user.User.UserName != userName {
		t.Fatalf("GetUser unexpected: %+v", user.User)
	}

	roleName := prefix + "-role"
	trust := `{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"},"Action":"sts:AssumeRole"}]}`
	_, err = client.CreateRole(ctx, &iam.CreateRoleInput{
		RoleName:                 aws.String(roleName),
		AssumeRolePolicyDocument: aws.String(trust),
	})
	if err != nil {
		t.Fatalf("CreateRole: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteRole(ctx, &iam.DeleteRoleInput{RoleName: aws.String(roleName)})
	})

	role, err := client.GetRole(ctx, &iam.GetRoleInput{RoleName: aws.String(roleName)})
	if err != nil {
		t.Fatalf("GetRole: %v", err)
	}
	if role.Role == nil || role.Role.Arn == nil || *role.Role.Arn == "" {
		t.Fatalf("GetRole missing ARN: %+v", role.Role)
	}

	_, err = client.DeleteRole(ctx, &iam.DeleteRoleInput{RoleName: aws.String(roleName)})
	if err != nil {
		t.Fatalf("DeleteRole: %v", err)
	}
	_, err = client.DeleteUser(ctx, &iam.DeleteUserInput{UserName: aws.String(userName)})
	if err != nil {
		t.Fatalf("DeleteUser: %v", err)
	}
}

func TestIAMAccessKeyLastUsedAndCredentialReport(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newIAM(t, cfg)
	ctx := context.Background()
	userName := uniquePrefix(t) + "-lastused"

	_, err := client.CreateUser(ctx, &iam.CreateUserInput{UserName: aws.String(userName)})
	if err != nil {
		t.Fatalf("CreateUser: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteUser(ctx, &iam.DeleteUserInput{UserName: aws.String(userName)})
	})

	keyOut, err := client.CreateAccessKey(ctx, &iam.CreateAccessKeyInput{UserName: aws.String(userName)})
	if err != nil {
		t.Fatalf("CreateAccessKey: %v", err)
	}
	if keyOut.AccessKey == nil || keyOut.AccessKey.AccessKeyId == nil || keyOut.AccessKey.SecretAccessKey == nil {
		t.Fatalf("CreateAccessKey missing key: %+v", keyOut.AccessKey)
	}
	accessKeyID := *keyOut.AccessKey.AccessKeyId
	secret := *keyOut.AccessKey.SecretAccessKey
	t.Cleanup(func() {
		_, _ = client.DeleteAccessKey(ctx, &iam.DeleteAccessKeyInput{
			UserName:    aws.String(userName),
			AccessKeyId: aws.String(accessKeyID),
		})
	})

	userCfg := cfg.Copy()
	userCfg.Credentials = credentials.NewStaticCredentialsProvider(accessKeyID, secret, "")
	userSTS := sts.NewFromConfig(userCfg, func(o *sts.Options) {
		o.BaseEndpoint = aws.String(endpoint())
	})
	if _, err := userSTS.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{}); err != nil {
		t.Fatalf("GetCallerIdentity with user key: %v", err)
	}

	lastUsed, err := client.GetAccessKeyLastUsed(ctx, &iam.GetAccessKeyLastUsedInput{
		AccessKeyId: aws.String(accessKeyID),
	})
	if err != nil {
		t.Fatalf("GetAccessKeyLastUsed: %v", err)
	}
	if lastUsed.UserName == nil || *lastUsed.UserName != userName {
		t.Fatalf("UserName=%v want %s", lastUsed.UserName, userName)
	}
	if lastUsed.AccessKeyLastUsed == nil {
		t.Fatal("AccessKeyLastUsed missing")
	}
	if aws.ToString(lastUsed.AccessKeyLastUsed.ServiceName) != "sts" {
		t.Fatalf("ServiceName=%q want sts", aws.ToString(lastUsed.AccessKeyLastUsed.ServiceName))
	}
	region := envOr("AWS_DEFAULT_REGION", "us-east-1")
	if aws.ToString(lastUsed.AccessKeyLastUsed.Region) != region {
		t.Fatalf("Region=%q want %s", aws.ToString(lastUsed.AccessKeyLastUsed.Region), region)
	}

	gen, err := client.GenerateCredentialReport(ctx, &iam.GenerateCredentialReportInput{})
	if err != nil {
		t.Fatalf("GenerateCredentialReport: %v", err)
	}
	if gen.State == "" || !strings.EqualFold(string(gen.State), "COMPLETE") {
		t.Fatalf("GenerateCredentialReport State=%q", gen.State)
	}

	report, err := client.GetCredentialReport(ctx, &iam.GetCredentialReportInput{})
	if err != nil {
		t.Fatalf("GetCredentialReport: %v", err)
	}
	if len(report.Content) == 0 {
		t.Fatal("GetCredentialReport empty Content")
	}
	rows, err := csv.NewReader(strings.NewReader(string(report.Content))).ReadAll()
	if err != nil {
		t.Fatalf("parse credential report CSV: %v", err)
	}
	found := false
	for _, row := range rows[1:] {
		if len(row) > 0 && row[0] == userName {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("credential report missing user %s: %s", userName, report.Content)
	}
}
